File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.java
class Solution {
    private void bellmanFord(long[][] arr, int s) {
        int V = arr.length;
        long[] dist = new long[V];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[s] = 0;
        for (int i = 0; i < V - 1; i++) {
            for (int u = 0; u < V; u++) {
                for (int v = 0; v < V; v++) {
                    if (dist[u] != Integer.MAX_VALUE && dist[u] + arr[u][v] < dist[v]) {
                        dist[v] = dist[u] + arr[u][v];
                    }
                }
            }
        }
        for (int i = 0; i < V; i++) {
            arr[s][i] = dist[i];
        }
    }
    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        long[][] arr = new long[26][26];
        for (int i = 0; i < 26; i++) {
            Arrays.fill(arr[i], Integer.MAX_VALUE);
        }
        int n = original.length;
        for (int i = 0; i < n; i++) {
            int u = original[i] - 'a';
            int v = changed[i] - 'a';
            arr[u][v] = Math.min((long)(cost[i]), arr[u][v]);
        }
        for (int i = 0; i < 26; i++) {
            arr[i][i] = 0;
        }
        for (int i = 0; i < 26; i++) {
            bellmanFord(arr, i);
        }
        n = target.length();
        long ans = 0;
        for (int i = 0; i < n; i++) {
            int u = source.charAt(i) - 'a';
            int v = target.charAt(i) - 'a';
            if (arr[u][v] == Integer.MAX_VALUE) {
                return -1;
            }
            ans += arr[u][v];
        }
        return ans;
    }
}










// class Solution {
//     private void floydWarshall(long[][] arr) {
//         long val = Integer.MAX_VALUE;
//         for (int via = 0; via < 26; via++) {
//             for (int i = 0; i < 26; i++) {
//                 for (int j = 0; j < 26; j++) {
//                     if (arr[i][via] < val && arr[via][j] < val) {
//                         arr[i][j] = Math.min(arr[i][j], arr[i][via] + arr[via][j]);
//                     }
//                 }
//             }
//         }
//     }
//     public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
//         long[][] arr = new long[26][26];
//         for (int i = 0; i < 26; i++) {
//             Arrays.fill(arr[i], Integer.MAX_VALUE);
//         }
//         int n = original.length;
//         for (int i = 0; i < n; i++) {
//             int u = original[i] - 'a';
//             int v = changed[i] - 'a';
//             arr[u][v] = Math.min((long)(cost[i]), arr[u][v]);
//         }
//         for (int i = 0; i < 26; i++) {
//             arr[i][i] = 0;
//         }
//         floydWarshall(arr);
//         n = target.length();
//         long ans = 0;
//         for (int i = 0; i < n; i++) {
//             int u = source.charAt(i) - 'a';
//             int v = target.charAt(i) - 'a';
//             if (arr[u][v] == Integer.MAX_VALUE) {
//                 return -1;
//             }
//             ans += arr[u][v];
//         }
//         return ans;
//     }
// }

File: LeetCode/2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.py
class Solution:
    def dijkstra(self, s, adj, arr):
        pq = []
        dist = [sys.maxsize] * 26
        dist[ord(s) - ord('a')] = 0
        heapq.heappush(pq, (0, s))
        
        while pq:
            d, n = heapq.heappop(pq)
            
            if d > dist[ord(n) - ord('a')]:
                continue
            
            for node, cost in adj[n]:
                if dist[ord(node) - ord('a')] > d + cost:
                    dist[ord(node) - ord('a')] = d + cost
                    heapq.heappush(pq, (d + cost, node))
        
        for i in range(26):
            arr[ord(s) - ord('a')][i] = dist[i]


    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        adj = defaultdict(list)
        for i in range(len(original)):
            adj[original[i]].append((changed[i], cost[i]))
        
        arr = [[sys.maxsize] * 26 for _ in range(26)]
        
        for c in range(26):
            self.dijkstra(chr(ord('a') + c), adj, arr)
        
        ans = 0
        n = len(source)
        
        for i in range(n):
            u = ord(source[i]) - ord('a')
            v = ord(target[i]) - ord('a')
            if arr[u][v] == sys.maxsize:
                return -1
            ans += arr[u][v]
        
        return ans











# class Solution:
#     def bellmanFord(self, arr, s):
#         V = len(arr)
#         dist = [sys.maxsize] * V
#         dist[s] = 0
#         for i in range(V-1):
#             for u in range(V):
#                 for v in range(V):
#                     if dist[u] != sys.maxsize and dist[u] + arr[u][v] < dist[v]:
#                         dist[v] = dist[u] + arr[u][v]
#         for i in range(V):
#             arr[s][i] = dist[i]

#     def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
#         arr = [[sys.maxsize for _ in range(26)] for _ in range(26)]
#         n = len(original)
#         for i in range(n):
#             u = ord(original[i]) - ord('a')
#             v = ord(changed[i])  - ord('a')
#             arr[u][v] = min(cost[i], arr[u][v])

#         for i in range(26):
#             arr[i][i] = 0
        
#         for i in range(26):
#             self.bellmanFord(arr, i)

#         n = len(target)

#         ans = 0

#         for i in range(n):
#             u = ord(source[i]) - ord('a')
#             v = ord(target[i]) - ord('a')
#             if arr[u][v] == sys.maxsize:
#                 return -1
#             ans += arr[u][v]
        
#         return ans










# class Solution:
#     def floydWarshall(self, arr):
#         for via in range(26):
#             for i in range(26):
#                 for j in range(26):
#                     arr[i][j] = min(arr[i][j], arr[i][via] + arr[via][j])

#     def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
#         arr = [[sys.maxsize for _ in range(26)] for _ in range(26)]
#         n = len(original)
#         for i in range(n):
#             u = ord(original[i]) - ord('a')
#             v = ord(changed[i])  - ord('a')
#             arr[u][v] = min(cost[i], arr[u][v])

#         for i in range(26):
#             arr[i][i] = 0
        
#         self.floydWarshall(arr)

#         n = len(target)

#         ans = 0

#         for i in range(n):
#             u = ord(source[i]) - ord('a')
#             v = ord(target[i]) - ord('a')
#             if arr[u][v] == sys.maxsize:
#                 return -1
#             ans += arr[u][v]
        
#         return ans